File: android/jsbridge/src/main/java/net/kibotu/jsbridge/commands/BridgeCommand.kt
package net.kibotu.jsbridge.commands

/**
 * Contract for implementing a single bridge command handler (Strategy Pattern).
 *
 * **Why this interface:**
 * Enables adding new bridge capabilities without modifying existing code (Open-Closed Principle).
 * Each command is self-contained: its own parsing, validation, execution, and error handling.
 *
 * **Why one action per command:**
 * Enforces Single Responsibility - each command does one thing well. Makes commands:
 * - Easy to test in isolation
 * - Easy to understand (no complex routing within command)
 * - Easy to maintain (changes to one action don't affect others)
 * - Easy to reuse across different bridge implementations
 *
 * **Why suspend:**
 * Commands often perform async operations (network, storage, UI). Suspend functions
 * allow efficient non-blocking execution without callback hell.
 *
 * **Design philosophy:**
 * Commands should be stateless, reusable, and independently testable. All state
 * comes from `content` parameter or injected dependencies.
 */
interface BridgeCommand {

    /**
     * The action identifier this command handles (e.g., "deviceInfo", "showToast").
     *
     * **Why string identifier:**
     * Matches web API contract where JavaScript sends action names as strings.
     * Must be unique across all registered commands.
     */
    val action: String

    /**
     * Executes the bridge command with provided parameters.
     *
     * **Why nullable return:**
     * - Non-null: Request-response commands that return data to web
     * - Null: Fire-and-forget commands (tracking, haptic) that don't need responses
     *
     * **Why nullable content:**
     * Some commands need no parameters (deviceInfo), others require complex objects.
     * Implementation is responsible for parsing and validating content.
     *
     * **Error handling:**
     * Commands should catch their own errors and return error response objects.
     * Throwing exceptions should be rare (only for unexpected/unrecoverable errors).
     *
     * @param content Optional parameters from web, typically JSONObject or JSON-serializable data
     * @return Response object for web (success/error) or null for fire-and-forget
     */
    suspend fun handle(content: Any?): Any?
}

File: ios/JSBridge/Sources/JSBridge/BridgeCommand.swift
import Foundation

/// Contract for implementing a single bridge command (Strategy Pattern).
///
/// **Why this pattern?**
/// - **Extensibility**: New commands can be added without modifying the core bridge logic
/// - **Single Responsibility**: Each command focuses on one specific action
/// - **Testability**: Commands can be tested in isolation without requiring a full WebView setup
/// - **Type Safety**: Ensures all commands conform to the same contract
///
/// **Design Decision:**
/// Uses the Strategy pattern to allow runtime registration of commands. This enables:
/// - Feature teams to add new commands independently
/// - Conditional command registration based on app configuration
/// - Easy mocking and testing of individual commands
public protocol BridgeCommand: Sendable {
    /// The action identifier this command handles (e.g., "deviceInfo", "showToast")
    var action: String { get }

    /// Handle the command with the given content.
    ///
    /// Throw `BridgeError` for expected failures (invalid parameters, missing data).
    /// Return nil for fire-and-forget commands that don't need a response.
    ///
    /// - Parameter content: Optional dictionary of parameters from JavaScript
    /// - Returns: Response dictionary for web, or nil for fire-and-forget
    @MainActor
    func handle(content: [String: Any]?) async throws -> [String: Any]?
}

/// Errors that can occur during bridge command handling
public enum BridgeError: Error, Sendable {
    case invalidMessage
    case unknownAction(String)
    case invalidParameter(String)
    case internalError(String)
    case unsupportedVersion(Int)
    
    public var code: String {
        switch self {
        case .invalidMessage:
            return "INVALID_MESSAGE"
        case .unknownAction:
            return "UNKNOWN_ACTION"
        case .invalidParameter:
            return "INVALID_PARAMETER"
        case .internalError:
            return "INTERNAL_ERROR"
        case .unsupportedVersion:
            return "UNSUPPORTED_VERSION"
        }
    }
    
    public var message: String {
        switch self {
        case .invalidMessage:
            return "Invalid message format"
        case .unknownAction(let action):
            return "Unknown action: \(action)"
        case .invalidParameter(let param):
            return "Invalid or missing parameter: \(param)"
        case .internalError(let msg):
            return "Internal error: \(msg)"
        case .unsupportedVersion(let version):
            return "Unsupported schema version: \(version)"
        }
    }
}
